app/controllers/bookmarks_controller.rb

```diff
@@ -0,0 +1,97 @@
+class BookmarksController < ApplicationController
+  require_login!
+  before_action :set_bookmark, only: [ :show, :edit, :update, :destroy ]
+
+  # GET /bookmarks
+  # GET /bookmarks.json
+  def index
+    @bookmarks = policy_scope(Bookmark).page params[:page]
+  end
+
+  # GET /bookmarks/new
+  def new
+    @bookmark = current_user.bookmarks.new
+    @bookmark.webpage = Webpage.new
+
+    authorize @bookmark
+  end
+
+  # POST /bookmarks
+  # POST /bookmarks.json
+  def create
+    @bookmark = current_user.bookmarks.new(bookmark_params)
+
+    authorize @bookmark
+
+    respond_to do |format|
+      if @bookmark.save
+        format.html { redirect_to @bookmark, notice: "Bookmark was successfully created." }
+        format.json { render :show, status: :created, location: @bookmark }
+      else
+        format.html { render :new }
+        format.json { render json: @bookmark.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # GET /bookmarks/1
+  # GET /bookmarks/1.json
+  def show
+    respond_to do |format|
+      format.html { render :show }
+      format.json { render :show, status: :ok }
+    end
+  end
+
+  # GET /bookmarks/1/edit
+  def edit
+  end
+
+  # PATCH/PUT /bookmarks/1
+  # PATCH/PUT /bookmarks/1.json
+  def update
+    respond_to do |format|
+      if @bookmark.update(bookmark_params)
+        format.html { redirect_to @bookmark, notice: "Bookmark was successfully updated." }
+        format.json { render :show, status: :ok, location: @bookmark }
+      else
+        format.html { render :edit }
+        format.json { render json: @bookmark.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  # DELETE /bookmarks/1
+  # DELETE /bookmarks/1.json
+  def destroy
+    respond_to do |format|
+      if @bookmark.destroy
+        format.html { redirect_to bookmarks_url, notice: "Bookmark was successfully deleted." }
+        format.json { head :no_content }
+      else
+        format.html { render :new }
+        format.json { render json: @bookmark.errors, status: :unprocessable_entity }
+      end
+    end
+  end
+
+  private
+
+  def set_bookmark
+    @bookmark = current_user.bookmarks.find(params[:id])
+    authorize @bookmark
+  end
+
+  def bookmark_params
+    permitted_attributes(@bookmark || Bookmark).tap do |obj|
+      tags = params.dig(:bookmark).fetch(:tags, []).map(&:strip).reject(&:empty?).map do |tag|
+        Tag.find_or_create_by label: tag
```

Comment:
What happens when javascript hiccups and sends this request twice and the database rejects the tag because it's non-unique?

Reply:
but its a find_or_create O_o?

Reply:
> Please note *this method is not atomic*, it runs first a SELECT, and if there are no results an INSERT is attempted. If there are other threads or processes there is a race condition between both calls and it could be the case that you end up with two similar records.
>
>Whether that is a problem or not depends on the logic of the application, but in the particular case in which rows have a UNIQUE constraint an exception may be raised

http://api.rubyonrails.org/classes/ActiveRecord/Relation.html#method-i-find_or_create_by

Reply:
Welp. TIL but yolo. I'm not sure if I'm happy with this pattern yet anyways though, so it might not be around long enough to encounter an issue.